### PR TITLE
fix #13945. support format on paste

### LINKED ISSUE #13945: Support format on paste
We should add support for format on paste. Idea is to simply invoke selection formatters when paste has happened.

### PR DESCRIPTION
**Pull request by rebornix**
Invoke format selection immediately after pasting event.

### CODE REVIEW

File: src/vs/editor/contrib/format/common/formatActions.ts [hunk 1]
```diff
@@ -235,6 +236,33 @@ export class FormatSelectionAction extends AbstractFormatAction {
 	}
 }
 
+@editorAction
+export class FormatOnPasteAction extends AbstractFormatAction {
+	constructor() {
+		super({
+			id: 'editor.action.formatOnPaste',
+			label: nls.localize('formatOnPaste.label', "Format on paste"),
+			alias: 'Format on paste',
```
Comment (jrieken) [inline src/vs/editor/contrib/format/common/formatActions.ts]: Maybe 'Paste and Format'?

File: src/vs/editor/contrib/format/common/formatActions.ts [hunk 2]
```diff
@@ -235,6 +236,33 @@ export class FormatSelectionAction extends AbstractFormatAction {
...
+			precondition: EditorContextKeys.Writable
```
Comment (jrieken) [inline src/vs/editor/contrib/format/common/formatActions.ts]: We have context key which knows if for the current document there is someone that can format selections: `ModeContextKeys.hasDocumentSelectionFormattingProvider` (see [`FormatSelection`](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/format/common/formatActions.ts#L219) as a sample). With that the menus will automatically update.

Comment (jrieken) [inline src/vs/editor/contrib/format/common/formatActions.ts]: Also register a keybinding and a menu item for this

**Review by jrieken — COMMENTED** [review]
Interesting approach. I was always thinking about a paste-listener and a setting but I actually like this. Left some nit comments, mostly around keybindings/menus

**Comment by rebornix** [comment]
@jrieken addressed your comments. Bind key chord `cmd+k cmd+v` to this action as there aren't choices for simple keybindings.

**Comment by jrieken** [comment]
On a second thought I am not so sure anymore if a separate action like 'Format and Paste' is the best solution. These are my concerns

* Having two actions I wonder if the usage patterns would become confusing. I might be a formatter-fan and bind `cmd+v` to 'paste&format' just to learn that paste is defunct for languages that don't have a formatter. That would force users to know if a formatter is available and then use a different keybinding.
* We already have settings like `formatOnType` and `formatOnSave` and I believe it would feel more natural to also have `formatOnPaste`

**Comment by rebornix** [comment]
@jrieken while implementing the command I have a feeling that ppl might prefer it as an option as well, I like `formatOnPaste`. I've already updated the code and follow the way we implement `formatOnType`.

**Comment by jrieken** [comment]
Code looks good to me. Waiting for @alexandrudima to review

File: src/vs/editor/common/commonCodeEditor.ts [hunk 3]
```diff
@@ -588,6 +589,25 @@ export abstract class CommonCodeEditor extends EventEmitter implements editorCom
...
+		if (handlerId === editorCommon.Handler.Paste) {
+			if (!this.cursor || typeof payload.text !== 'string' || payload.text.length === 0) {
+				// nothing to do
+				return;
+			}
+			const startPosition = this.cursor.getSelection().getStartPosition();
+			this.cursor.trigger(source, handlerId, payload);
+			const endPosition = this.cursor.getSelection().getStartPosition();
+			if (source === 'keyboard') {
+				this.emit(editorCommon.EventType.DidPaste, {
```
Comment (alexandrudima) [inline src/vs/editor/common/commonCodeEditor.ts]: I would prefer for the editor to always give out rich objects (i.e. `Range` and not `IRange`). It does so for all non-serializable events.

File: src/vs/editor/common/editorCommon.ts [hunk 4]
```diff
@@ -3801,6 +3810,13 @@ export interface ICommonCodeEditor extends IEditor {
 	onDidType(listener: (text: string) => void): IDisposable;
 
 	/**
+	 * An event emitted when users paste text in the editor.
+	 * @event
+	 * @internal
+	 */
+	onDidPaste(listener: (range: Range) => void): IDisposable;
```
Comment (alexandrudima) [inline src/vs/editor/common/editorCommon.ts]: Here we say we give out a `Range`, but the implementation gives out an `IRange`. Please adjust the implementation.

File: src/vs/editor/contrib/format/common/formatActions.ts [hunk 5]
```diff
@@ -150,6 +151,87 @@ class FormatOnType implements editorCommon.IEditorContribution {
...
+		if (!this.editor.getConfiguration().contribInfo.formatOnPaste) {
+			return;
+		}
+
+		// no model
+		if (!this.editor.getModel()) {
+			return;
+		}
+
+		var model = this.editor.getModel();
```
Comment (alexandrudima) [inline src/vs/editor/contrib/format/common/formatActions.ts]: nitpick: prefer let over var (same two lines below)

File: src/vs/editor/contrib/format/common/formatActions.ts [hunk 6]
```diff
@@ -150,6 +151,87 @@ class FormatOnType implements editorCommon.IEditorContribution {
...
+
+		// no support
+		var [support] = OnTypeFormattingEditProviderRegistry.ordered(model);
+		if (!support || !support.autoFormatTriggerCharacters) {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/format/common/formatActions.ts]: No need for the support to have autoFormatTriggerCharacters. Perhaps a copy-paste error?

File: src/vs/editor/contrib/format/common/formatActions.ts [hunk 7]
```diff
@@ -150,6 +151,87 @@ class FormatOnType implements editorCommon.IEditorContribution {
...
+
+		getDocumentRangeFormattingEdits(model, range, { tabSize, insertSpaces }).then(edits => {
+			return this.workerService.computeMoreMinimalEdits(model.uri, edits, []);
+		}).then(edits => {
+			if (!state.validate(this.editor) || isFalsyOrEmpty(edits)) {
+				return;
+			}
+			const command = new EditOperationsCommand(edits, this.editor.getSelection());
+			this.editor.executeCommand(this.getId(), command);
+			this.editor.focus();
```
Comment (alexandrudima) [inline src/vs/editor/contrib/format/common/formatActions.ts]: No need to focus here, since we are not moving the focus anywhere.

**Review by alexdima — CHANGES_REQUESTED** [review]
Looks good overall, just a few details left.

File: src/vs/editor/contrib/format/common/formatActions.ts (hunk 6 above)
Comment (alexandrudima) [inline src/vs/editor/contrib/format/common/formatActions.ts]: Now that I look at it, we should not read the OnTypeFormattingEditProvierRegistry, but the DocumentRangeFormatting bla bla Registry.

Comment (rebornix) [inline src/vs/editor/contrib/format/common/formatActions.ts]: Yup it's a stupid copy-paste error, I made changes to one place but forgot it here.

**Comment by rebornix** [comment]
@alexandrudima comments addressed.

**Review by alexdima — APPROVED** [review]

**Comment by alexandrudima** [comment]
Brilliant!

**Comment by alexandrudima** [comment]
@rebornix remember to create a test plan item for this

**Comment by jrieken** [comment]
@rebornix @alexandrudima i have done that

**Pull request merged**